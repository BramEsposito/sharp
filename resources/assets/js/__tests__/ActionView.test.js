import Notifications from 'vue-notification';
import { shallowMount, mount, createLocalVue, config } from '@vue/test-utils';
import ActionView from '../components/ActionView.vue';
import Vue from 'vue';

config.stubs['transition-group'] = false;

describe('action-view', ()=>{

    const ModalStub = {
        name:'TestModal',
        template:'<div><slot/></div>'
    };

    const localVue = createLocalVue();
    localVue.use(Notifications);

    function notify(wrapper, ...args) {
        wrapper.vm.$notify(...args);
        // remove notification global data-id to preserve unit
        wrapper.findAll('[data-id]').wrappers.forEach(wrapper=>wrapper.element.removeAttribute('data-id'));
    }

    function createWrapper(options={}) {
        return shallowMount(ActionView, {
            ...options,
            propsData: {
                context: 'form',
                ...options.propsData
            },
            stubs: {
                'SharpModal': ModalStub
            },
            localVue
        })
    }


    test('can mount ActionView', ()=>{
        expect(createWrapper().html()).toMatchSnapshot();
    });

    test('can mount ActionView with notification', ()=>{
        const wrapper = createWrapper({ notifications: true });
        notify(wrapper, { title:'TITLE', text:'MESSAGE', type:'TYPE' });
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('can mount ActionView with closed notification', ()=>{
        const wrapper = createWrapper({ notifications:true });
        notify(wrapper, 'message');
        wrapper.find('[data-test=close-notification]').trigger('click');
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('can mount ActionView with error page', ()=>{
        const wrapper = createWrapper({
            data: {
                showErrorPage: true,
                errorPageData: {
                    status: 404,
                    message: 'ERROR MESSAGE'
                }
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('can mount with action bar', ()=>{
        const BarComponent = { name:'TestBarComponent', template:'<div>BAR COMPONENT</div>' };
        const wrapper = createWrapper({
            computed: {
                barComp:()=>BarComponent
            }
        });
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('can mount with multiple modals', ()=>{
        const wrapper = createWrapper();
        wrapper.vm.showMainModal({ text: 'Modal 1' });
        wrapper.vm.showMainModal({ text: 'Modal 2' });
        expect(wrapper.html()).toMatchSnapshot();
    });

    xtest('handle main modal events', ()=>{
        const wrapper = createWrapper();
        const modalOptions = { text: 'Modal 1', okCallback:jest.fn(), hiddenCallback:jest.fn() };
        wrapper.vm.showMainModal(modalOptions);
        
        let modal = wrapper.find(ModalStub);

        modal.vm.$emit('ok');
        expect(modalOptions.okCallback).toHaveBeenCalled();
    });
});