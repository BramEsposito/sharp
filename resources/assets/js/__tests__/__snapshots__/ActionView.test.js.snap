// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action-view can mount ActionView 1`] = `
<div class="SharpActionView">
	<!---->
	<div class="container">
		<!---->
	</div>
</div>
`;

exports[`action-view can mount ActionView with closed notification 1`] = `
<div class="SharpActionView">
	<!---->
	<div class="container">
		<div class="notifications" style="width: 300px; right: 0px;">
			<span>
				<div class="notification-wrapper slideRight-leave slideRight-leave-active">
					<div role="alert" data-test="notification" class="SharpToastNotification SharpToastNotification--info">
						<div class="SharpToastNotification__details">
							<h3 class="SharpToastNotification__title mb-2"></h3>
							<p class="SharpToastNotification__caption">message</p>
						</div>
						<button type="button" data-test="close-notification" class="SharpToastNotification__close-button">
							<svg aria-label="close" width="10" height="10" viewBox="0 0 10 10" fill-rule="evenodd" class="SharpToastNotification__icon">
								<path d="M9.8 8.6L8.4 10 5 6.4 1.4 10 0 8.6 3.6 5 .1 1.4 1.5 0 5 3.6 8.6 0 10 1.4 6.4 5z"></path>
							</svg>
						</button>
					</div>
				</div>
			</span>
		</div>
	</div>
</div>
`;

exports[`action-view can mount ActionView with error page 1`] = `
<div class="SharpActionView">
	<!---->
	<div class="container">
		<h1>Error 404</h1>
		<p>ERROR MESSAGE</p>
	</div>
</div>
`;

exports[`action-view can mount ActionView with notification 1`] = `
<div class="SharpActionView">
	<!---->
	<div class="container">
		<div class="notifications" style="width: 300px; right: 0px;">
			<span>
				<div class="notification-wrapper slideRight-enter slideRight-enter-active">
					<div role="alert" data-test="notification" class="SharpToastNotification SharpToastNotification--TYPE">
						<div class="SharpToastNotification__details">
							<h3 class="SharpToastNotification__title mb-2">TITLE</h3>
							<p class="SharpToastNotification__caption">MESSAGE</p>
						</div>
						<button type="button" data-test="close-notification" class="SharpToastNotification__close-button">
							<svg aria-label="close" width="10" height="10" viewBox="0 0 10 10" fill-rule="evenodd" class="SharpToastNotification__icon">
								<path d="M9.8 8.6L8.4 10 5 6.4 1.4 10 0 8.6 3.6 5 .1 1.4 1.5 0 5 3.6 8.6 0 10 1.4 6.4 5z"></path>
							</svg>
						</button>
					</div>
				</div>
			</span>
		</div>
	</div>
</div>
`;

exports[`action-view can mount with action bar 1`] = `
<div class="SharpActionView">
	<div>BAR COMPONENT</div>
	<div class="container">
		<div class="notifications" style="width: 300px; right: 0px;">
			<span></span>
		</div>
	</div>
</div>
`;

exports[`action-view can mount with multiple modals 1`] = `
<div class="SharpActionView">
	<!---->
	<div class="container">
		<div class="notifications" style="width: 300px; right: 0px;">
			<span></span>
		</div>
		<div visible="visible">
			Modal 1
		</div>
		<div visible="visible">
			Modal 2
		</div>
	</div>
</div>
`;
